File: src/main/scala/laoflch/debezium/connector/informix/InformixSnapshotChangeEventSource.scala
package laoflch.debezium.connector.informix

import io.debezium.pipeline.EventDispatcher
import io.debezium.pipeline.source.AbstractSnapshotChangeEventSource
import io.debezium.pipeline.source.spi.{ChangeEventSource, SnapshotProgressListener}
import io.debezium.pipeline.spi.OffsetContext
import io.debezium.pipeline.txmetadata.TransactionContext
import io.debezium.relational.{RelationalSnapshotChangeEventSource, Table, TableId}
import io.debezium.schema.SchemaChangeEvent
import io.debezium.schema.SchemaChangeEvent.SchemaChangeEventType
import io.debezium.util.Clock
import laoflch.debezium.connector.informix.InformixConnectorConfig.SnapshotIsolationMode
import org.slf4j.LoggerFactory

import java.sql.{Connection, SQLException, Savepoint}
import java.util
import java.util.{Optional, Set}
import scala.jdk.CollectionConverters


object InformixSnapshotChangeEventSource {

  @throws[SQLException]
  class InformixSnapshotContext(catalogName: String) extends RelationalSnapshotChangeEventSource.RelationalSnapshotContext(catalogName) {
    var isolationLevelBeforeStart = 0
    var preSchemaSnapshotSavepoint: Savepoint = null
  }
}


class InformixSnapshotChangeEventSource(connectorConfig: InformixConnectorConfig, previousOffset: InformixOffsetContext, jdbcConnection: InformixConnection, schema: InformixDatabaseSchema,
                                        dispatcher: EventDispatcher[TableId], clock: Clock, snapshotProgressListener: SnapshotProgressListener)
  extends RelationalSnapshotChangeEventSource(connectorConfig, previousOffset, jdbcConnection, schema,
    dispatcher, clock, snapshotProgressListener) {

  private val LOGGER = LoggerFactory.getLogger(classOf[InformixSnapshotChangeEventSource])

  /*private val connectorConfig = connectorConfig
  private val jdbcConnection = jdbcConnection*/


  override protected def getSnapshottingTask(previousOffset: OffsetContext): AbstractSnapshotChangeEventSource.SnapshottingTask = {
    var snapshotSchema = true
    var snapshotData = true
    // found a previous offset and the earlier snapshot has completed
    if (previousOffset != null && !previousOffset.isSnapshotRunning) {
      LOGGER.info("A previous offset indicating a completed snapshot has been found. Neither schema nor data will be snapshotted.")
      snapshotSchema = false
      snapshotData = false
    }
    else {
      LOGGER.info("No previous offset has been found")
      if (connectorConfig.getSnapshotMode.includeData)
        LOGGER.info("According to the connector configuration both schema and data will be snapshotted")
      else
        LOGGER.info("According to the connector configuration only schema will be snapshotted")
      snapshotData = connectorConfig.getSnapshotMode.includeData
    }
    new AbstractSnapshotChangeEventSource.SnapshottingTask(snapshotSchema, snapshotData)
  }

  @throws[Exception]
  override protected def prepare(context: ChangeEventSource.ChangeEventSourceContext) = new InformixSnapshotChangeEventSource.InformixSnapshotContext(jdbcConnection.getRealDatabaseName)

  @throws[Exception]
  override protected def connectionCreated(snapshotContext: RelationalSnapshotChangeEventSource.RelationalSnapshotContext): Unit = {
    snapshotContext.asInstanceOf[InformixSnapshotChangeEventSource.InformixSnapshotContext].isolationLevelBeforeStart = jdbcConnection.connection.getTransactionIsolation
  }

  @throws[Exception]
  override protected def getAllTableIds(ctx: RelationalSnapshotChangeEventSource.RelationalSnapshotContext): Set[TableId] = {
    jdbcConnection.readTableNames(ctx.catalogName, null, null, Array[String]("TABLE"))
  }

  @throws[SQLException]
  @throws[InterruptedException]
  override protected def lockTablesForSchemaSnapshot(sourceContext: ChangeEventSource.ChangeEventSourceContext, snapshotContext: RelationalSnapshotChangeEventSource.RelationalSnapshotContext): Unit = {
    if (connectorConfig.getSnapshotIsolationMode == SnapshotIsolationMode.READ_UNCOMMITTED) {
      jdbcConnection.connection.setTransactionIsolation(Connection.TRANSACTION_READ_UNCOMMITTED)
      LOGGER.info("Schema locking was disabled in connector configuration")
    }
    else if (connectorConfig.getSnapshotIsolationMode == SnapshotIsolationMode.READ_COMMITTED) {
      jdbcConnection.connection.setTransactionIsolation(Connection.TRANSACTION_READ_COMMITTED)
      LOGGER.info("Schema locking was disabled in connector configuration")
    }
    else if ( /*(connectorConfig.getSnapshotIsolationMode == SnapshotIsolationMode.EXCLUSIVE) ||*/
      (connectorConfig.getSnapshotIsolationMode == SnapshotIsolationMode.REPEATABLE_READ)) {
      jdbcConnection.connection.setTransactionIsolation(Connection.TRANSACTION_REPEATABLE_READ)
      /*snapshotContext.asInstanceOf[InformixSnapshotChangeEventSource.InformixSnapshotContext].preSchemaSnapshotSavepoint =
        jdbcConnection.connection.setSavepoint("informix_schema_snapshot")
      LOGGER.info("Executing schema locking")
      try {
        val statement = jdbcConnection.connection.createStatement(ResultSet.TYPE_FORWARD_ONLY, ResultSet.CONCUR_READ_ONLY)
        try {
         // import scala.collection.JavaConversions._
          for (tableId <- CollectionConverters.SetHasAsScala(snapshotContext.capturedTables).asScala) {
            if (!sourceContext.isRunning) throw new InterruptedException("Interrupted while locking table " + tableId)
            LOGGER.info("Locking table {}", tableId)
            val query = String.format("SELECT * FROM %s.%s WHERE 0=1 WITH CS", tableId.schema, tableId.table)
            statement.executeQuery(query).close()
          }
         }finally
          {
            if (statement != null) statement.close()
          }
        }*/

    }
    else throw new IllegalStateException("Unknown locking mode specified.")
  }

  @throws[SQLException]
  override protected def releaseSchemaSnapshotLocks(snapshotContext: RelationalSnapshotChangeEventSource.RelationalSnapshotContext): Unit = { // Exclusive mode: locks should be kept until the end of transaction.
    // read_uncommitted mode; read_committed mode: no locks have been acquired.
    if (connectorConfig.getSnapshotIsolationMode eq SnapshotIsolationMode.REPEATABLE_READ) {
      //jdbcConnection.connection.rollback(snapshotContext.asInstanceOf[InformixSnapshotChangeEventSource.InformixSnapshotContext].preSchemaSnapshotSavepoint)
      LOGGER.info("Schema locks released.")
    }
  }

  @throws[Exception]
  override protected def determineSnapshotOffset(ctx: RelationalSnapshotChangeEventSource.RelationalSnapshotContext): Unit = {
    val offset = new InformixOffsetContext(connectorConfig, TxLogPosition.NULL, false, false, TransactionContext.load(new util.HashMap[String, Any]()))
    // offset.setCDCEngine(jdbcConnection.getCDCEngine())
    ctx.offset = offset
  }

  @throws[SQLException]
  @throws[InterruptedException]
  override protected def readTableStructure(sourceContext: ChangeEventSource.ChangeEventSourceContext,
                                            snapshotContext: RelationalSnapshotChangeEventSource.RelationalSnapshotContext): Unit = {

    LOGGER.info("snapshotContext.capturedTables={}", snapshotContext.capturedTables)
    val schemas = CollectionConverters.SetHasAsScala(snapshotContext.capturedTables).asScala.map[String]((x: TableId) => x.schema)
    LOGGER.info("readTableStructure, schema={}", schemas)

    // reading info only for the schemas we're interested in as per the set of captured tables;
    // while the passed table name filter alone would skip all non-included tables, reading the schema
    // would take much longer that way

    for (schema <- schemas) {
      if (!sourceContext.isRunning) throw new InterruptedException("Interrupted while reading structure of schema " + schema)
      LOGGER.info("Reading structure of schema '{}'", schema)
      jdbcConnection.readSchema(snapshotContext.tables,
        snapshotContext.catalogName,
        schema,
        connectorConfig.getTableFilters.dataCollectionFilter,
        null, false)
    }

    // add tables to offset use by cdc lsn

/*
    jdbcConnection.getCDCEngine().setTableColsMap(
      CollectionConverters.SetHasAsScala(snapshotContext.tables.tableIds()).asScala.map[(String, CDCTabeEntry)](
        x => (x.catalog() + ":" + x.schema() + ":" + x.table() ->
          CDCTabeEntry(x,
            CollectionConverters
              .ListHasAsScala(snapshotContext.tables.forTable(x.catalog(), x.schema(), x.table()).columns())
              .asScala
              .map[String](x => x.name()).toSeq)
          ))
        .toMap
    )
*/

    /* def extractColsFormTable(cols:java.util.List[Column]): Seq[String] ={
       CollectionConverters.ListHasAsScala(cols).asScala.map[String](x=>x.name()).toSeq
     }*/

    //snapshotContext.offset.asInstanceOf[InformixOffsetContext].getCDCEngine().watchTableAndCols=watchTableAndCols

  }

  @throws[SQLException]
  override protected def getCreateTableEvent(snapshotContext: RelationalSnapshotChangeEventSource.RelationalSnapshotContext,
                                             table: Table) =
    new SchemaChangeEvent(snapshotContext.offset.getPartition,
      snapshotContext.offset.getOffset,
      snapshotContext.catalogName,
      table.id.schema,
      null, table, SchemaChangeEventType.CREATE, true)

  override protected def complete(snapshotContext: AbstractSnapshotChangeEventSource.SnapshotContext): Unit = {
    try jdbcConnection.connection.setTransactionIsolation(
      snapshotContext.asInstanceOf[InformixSnapshotChangeEventSource.InformixSnapshotContext].isolationLevelBeforeStart
    )
    catch {
      case e: SQLException =>
        throw new RuntimeException("Failed to set transaction isolation level.", e)
    }
  }

  /**
   * Generate a valid Informix query string for the specified table
   *
   * @param tableId the table to generate a query for
   * @return a valid query string
   */
  override protected def getSnapshotSelect(snapshotContext: AbstractSnapshotChangeEventSource.SnapshotContext,
                                           tableId: TableId): Optional[String] =
    Optional.of(String.format("SELECT * FROM %s.%s", tableId.schema, tableId.table))

}

File: src/main/java/laoflch/debezium/connector/informix/InformixSnapshotChangeEventSource.java
/*
 * Copyright Debezium-Informix-Connector Authors.
 *
 * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
 */

package laoflch.debezium.connector.informix;

import java.sql.Connection;
import java.sql.SQLException;
import java.sql.Savepoint;
import java.util.HashMap;
import java.util.Optional;
import java.util.Set;
import java.util.stream.Collectors;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import io.debezium.pipeline.EventDispatcher;
import io.debezium.pipeline.source.spi.SnapshotProgressListener;
import io.debezium.pipeline.spi.OffsetContext;
import io.debezium.pipeline.txmetadata.TransactionContext;
import io.debezium.relational.RelationalSnapshotChangeEventSource;
import io.debezium.relational.Table;
import io.debezium.relational.TableId;
import io.debezium.schema.SchemaChangeEvent;
import io.debezium.schema.SchemaChangeEvent.SchemaChangeEventType;
import io.debezium.util.Clock;

import laoflch.debezium.connector.informix.InformixConnectorConfig.SnapshotIsolationMode;

public class InformixSnapshotChangeEventSource extends RelationalSnapshotChangeEventSource {

    private static final Logger LOGGER = LoggerFactory.getLogger(InformixSnapshotChangeEventSource.class);

    private final InformixConnectorConfig connectorConfig;
    private final InformixConnection jdbcConnection;

    public InformixSnapshotChangeEventSource(InformixConnectorConfig connectorConfig, InformixOffsetContext previousOffset, InformixConnection jdbcConnection,
                                             InformixDatabaseSchema schema,
                                             EventDispatcher<TableId> dispatcher, Clock clock, SnapshotProgressListener snapshotProgressListener) {
        super(connectorConfig, previousOffset, jdbcConnection, schema, dispatcher, clock, snapshotProgressListener);
        this.connectorConfig = connectorConfig;
        this.jdbcConnection = jdbcConnection;
    }

    @Override
    protected SnapshottingTask getSnapshottingTask(OffsetContext previousOffset) {
        boolean snapshotSchema = true;
        boolean snapshotData = true;

        // found a previous offset and the earlier snapshot has completed
        if (previousOffset != null && !previousOffset.isSnapshotRunning()) {
            LOGGER.info("A previous offset indicating a completed snapshot has been found. Neither schema nor data will be snapshotted.");
            snapshotSchema = false;
            snapshotData = false;
        }
        else {
            LOGGER.info("No previous offset has been found");
            if (connectorConfig.getSnapshotMode().includeData()) {
                LOGGER.info("According to the connector configuration both schema and data will be snapshotted");
            }
            else {
                LOGGER.info("According to the connector configuration only schema will be snapshotted");
            }
            snapshotData = connectorConfig.getSnapshotMode().includeData();
        }

        return new SnapshottingTask(snapshotSchema, snapshotData);
    }

    @Override
    protected SnapshotContext prepare(ChangeEventSourceContext context) throws Exception {
        return new InformixSnapshotChangeEventSource.InformixSnapshotContext(jdbcConnection.getRealDatabaseName());
    }

    @Override
    protected void connectionCreated(RelationalSnapshotContext snapshotContext) throws Exception {
        ((InformixSnapshotContext) snapshotContext).isolationLevelBeforeStart = jdbcConnection.connection().getTransactionIsolation();
    }

    @Override
    protected Set<TableId> getAllTableIds(RelationalSnapshotContext ctx) throws Exception {
        return jdbcConnection.readTableNames(null, null, null, new String[]{ "TABLE" });
    }

    @Override
    protected void lockTablesForSchemaSnapshot(ChangeEventSourceContext sourceContext, RelationalSnapshotContext snapshotContext)
            throws SQLException, InterruptedException {
        if (connectorConfig.getSnapshotIsolationMode() == SnapshotIsolationMode.READ_UNCOMMITTED) {
            jdbcConnection.connection().setTransactionIsolation(Connection.TRANSACTION_READ_UNCOMMITTED);
            LOGGER.info("Schema locking was disabled in connector configuration");
        }
        else if (connectorConfig.getSnapshotIsolationMode() == SnapshotIsolationMode.READ_COMMITTED) {
            jdbcConnection.connection().setTransactionIsolation(Connection.TRANSACTION_READ_COMMITTED);
            LOGGER.info("Schema locking was disabled in connector configuration");
        }
        else if (/* connectorConfig.getSnapshotIsolationMode() == SnapshotIsolationMode.EXCLUSIVE || */
        connectorConfig.getSnapshotIsolationMode() == SnapshotIsolationMode.REPEATABLE_READ) {
            jdbcConnection.connection().setTransactionIsolation(Connection.TRANSACTION_REPEATABLE_READ);
            // TODO: confirm these modifications
            /*
             * ((InformixSnapshotContext) snapshotContext).preSchemaSnapshotSavepoint = jdbcConnection.connection().setSavepoint("informix_schema_snapshot");
             * 
             * LOGGER.info("Executing schema locking");
             * try (Statement statement = jdbcConnection.connection().createStatement(ResultSet.TYPE_FORWARD_ONLY, ResultSet.CONCUR_READ_ONLY)) {
             * for (TableId tableId : snapshotContext.capturedTables) {
             * if (!sourceContext.isRunning()) {
             * throw new InterruptedException("Interrupted while locking table " + tableId);
             * }
             * 
             * LOGGER.info("Locking table {}", tableId);
             * String query = String.format("SELECT * FROM %s.%s WHERE 0=1 WITH CS", Db2ObjectNameQuoter.quoteNameIfNecessary(tableId.schema()),
             * Db2ObjectNameQuoter.quoteNameIfNecessary(tableId.table()));
             * statement.executeQuery(query).close();
             * }
             * }
             */
        }
        else {
            throw new IllegalStateException("Unknown locking mode specified.");
        }
    }

    @Override
    protected void releaseSchemaSnapshotLocks(RelationalSnapshotContext snapshotContext) throws SQLException {
        // Exclusive mode: locks should be kept until the end of transaction.
        // read_uncommitted mode; read_committed mode: no locks have been acquired.
        if (connectorConfig.getSnapshotIsolationMode() == SnapshotIsolationMode.REPEATABLE_READ) {
            // TODO: confirm this modification
            // jdbcConnection.connection().rollback(((InformixSnapshotContext) snapshotContext).preSchemaSnapshotSavepoint);
            LOGGER.info("Schema locks released.");
        }
    }

    @Override
    protected void determineSnapshotOffset(RelationalSnapshotContext ctx) throws Exception {
        // TODO: Check the following modification
        ctx.offset = new InformixOffsetContext(
                connectorConfig,
                // TxLogPosition.valueOf(jdbcConnection.getMaxLsn()),
                TxLogPosition.NULL,
                false,
                false,
                //
                TransactionContext.load(new HashMap<String, Object>()));
    }

    @Override
    protected void readTableStructure(ChangeEventSourceContext sourceContext, RelationalSnapshotContext snapshotContext) throws SQLException, InterruptedException {
        Set<String> schemas = snapshotContext.capturedTables.stream()
                .map(TableId::schema)
                .collect(Collectors.toSet());

        // reading info only for the schemas we're interested in as per the set of captured tables;
        // while the passed table name filter alone would skip all non-included tables, reading the schema
        // would take much longer that way
        for (String schema : schemas) {
            if (!sourceContext.isRunning()) {
                throw new InterruptedException("Interrupted while reading structure of schema " + schema);
            }

            LOGGER.info("Reading structure of schema '{}'", schema);

            jdbcConnection.readSchema(
                    snapshotContext.tables,
                    null,
                    schema,
                    connectorConfig.getTableFilters().dataCollectionFilter(),
                    null,
                    false);
        }
    }

    @Override
    protected SchemaChangeEvent getCreateTableEvent(RelationalSnapshotContext snapshotContext, Table table) throws SQLException {
        return new SchemaChangeEvent(
                snapshotContext.offset.getPartition(),
                snapshotContext.offset.getOffset(),
                snapshotContext.offset.getSourceInfo(),
                snapshotContext.catalogName,
                table.id().schema(),
                null,
                table,
                SchemaChangeEventType.CREATE, true);
    }

    @Override
    protected void complete(SnapshotContext snapshotContext) {
        try {
            jdbcConnection.connection().setTransactionIsolation(((InformixSnapshotContext) snapshotContext).isolationLevelBeforeStart);
        }
        catch (SQLException e) {
            throw new RuntimeException("Failed to set transaction isolation level.", e);
        }
    }

    /**
     * Generate a valid informix query string for the specified table
     *
     * @param tableId the table to generate a query for
     * @return a valid query string
     */
    @Override
    protected Optional<String> getSnapshotSelect(RelationalSnapshotContext snapshotContext, TableId tableId) {
        return Optional.of(String.format("SELECT * FROM %s.%s", tableId.schema(), tableId.table()));
    }

    /**
     * Mutable context which is populated in the course of snapshotting.
     */
    private static class InformixSnapshotContext extends RelationalSnapshotContext {

        private int isolationLevelBeforeStart;
        private Savepoint preSchemaSnapshotSavepoint;

        public InformixSnapshotContext(String catalogName) throws SQLException {
            super(catalogName);
        }
    }
}
